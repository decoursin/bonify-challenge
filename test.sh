#! /bin/bash

set -euo pipefail

host=${HOST:-localhost}
port=${PORT:-8080}
actual=${TEST_FILENAME:-test.actual.csv}
expected=${TEST_FILENAME:-test.expected.csv}

generate_post_data()
{
    cat <<EOF
{
    "Partner_Account": "${line[0]}",
    "Partner_BLZ": "${line[2]}",
    "Bank_Name": "${line[3]}",
    "Partner_Name": "${line[4]}",
    "Booking_Text": "${line[5]}",
    "Subject": "${line[6]}",
    "Booking_Date": "${line[7]}",
    "Transfer_Type": "${line[8]}",
    "Currency": "${line[9]}",
    "Amount": "${line[10]}"
}
EOF
}

## truncate actual testing file.
## and add header
echo "account_number;datetime;spending (10-day window)" > "$actual"

## truncate table for accurate calculations.
psql -U postgres -h localhost -c 'truncate table account;' "bonify"

## Send data to API.
IFS=";"
{
    read; # skip header
    while read -a line; do
        echo "body: $(generate_post_data line)"

        curl -X PUT \
             -H "Accept: application/json" \
             -H "Content-Type: application/json" \
             --data "$(generate_post_data line)" \
             "$host:$port/transaction"
    done
} < account.csv

## Assert expected == actual
DIFF=$(diff "$expected" "$actual") 
if [ "$DIFF" != "" ] 
then
    echo "FAILING TEST! $DIFF"
else
	echo "Test pass! Nice job."
fi
