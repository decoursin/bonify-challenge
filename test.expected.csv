account_number;datetime;spending (10-day window)
4711951501;2015-05-06T02:00+02:00;32.92
4711951501;2015-05-08T02:00+02:00;282.92
4711951501;2015-05-09T02:00+02:00;520.92
4711951501;2015-05-10T02:00+02:00;535.91
4711951501;2015-05-11T02:00+02:00;545.91
4711951501;2015-05-12T02:00+02:00;565.81
4711951501;2015-05-13T02:00+02:00;732.93
4711951501;2015-05-14T02:00+02:00;758.13
4711951501;2015-05-15T02:00+02:00;1511.33
4711951501;2015-05-16T02:00+02:00;1496.31
4987464748;2015-05-17T02:00+02:00;0
4711951501;2015-05-18T02:00+02:00;1546.31
4711951501;2015-05-19T02:00+02:00;1403.94
4711951501;2015-05-20T02:00+02:00;2588.95
4711951501;2015-05-21T02:00+02:00;2614.07
4711951501;2015-05-23T02:00+02:00;2505.65
4711951501;2015-05-24T02:00+02:00;2480.45
4711951501;2015-05-25T02:00+02:00;1863.05
4711951501;2015-05-26T02:00+02:00;2245.15
4711951501;2015-05-27T02:00+02:00;2379.27
4987464748;2015-05-28T02:00+02:00;0
4711951501;2015-05-29T02:00+02:00;2075.19
4711951501;2015-05-30T02:00+02:00;930.08
4711951501;2015-05-31T02:00+02:00;1194.96
4711951501;2015-06-01T02:00+02:00;1313.01
4711951501;2015-06-02T02:00+02:00;1324.66
4711951501;2015-06-03T02:00+02:00;1324.66
4711951501;2015-06-04T02:00+02:00;1229.87
4711951501;2015-06-09T02:00+02:00;628.56
4711951501;2015-06-11T02:00+02:00;210.51
4987464748;2015-06-12T02:00+02:00;0
4711951501;2015-06-13T02:00+02:00;350.26
4711951501;2015-06-14T02:00+02:00;339.15
4711951501;2015-06-15T02:00+02:00;397.07
4711951501;2015-06-16T02:00+02:00;627.07
4711951501;2015-06-17T02:00+02:00;637.07
4711951501;2015-06-18T02:00+02:00;1390.27
4711951501;2015-06-19T02:00+02:00;1336.98
4711951501;2015-06-20T02:00+02:00;1341.97
4711951501;2015-06-21T02:00+02:00;1360.47
4711951501;2015-06-22T02:00+02:00;1372.17
4711951501;2015-06-23T02:00+02:00;1175.09
4711951501;2015-06-24T02:00+02:00;1345.19
4711951501;2015-06-25T02:00+02:00;1587.27
4711951501;2015-06-26T02:00+02:00;1595.27
4711951501;2015-06-27T02:00+02:00;1600.26
4711951501;2015-06-28T02:00+02:00;1014.18
4711951501;2015-06-29T02:00+02:00;1008.12
4711951501;2015-06-30T02:00+02:00;1003.13
4711951501;2015-07-01T02:00+02:00;1009.83
4711951501;2015-07-02T02:00+02:00;1298.13
4711951501;2015-07-03T02:00+02:00;1283.11
4711951501;2015-07-04T02:00+02:00;1112.01
4711951501;2015-07-05T02:00+02:00;862.11
4711951501;2015-07-07T02:00+02:00;839.12
4711951501;2015-07-08T02:00+02:00;701.90
4711951501;2015-07-09T02:00+02:00;739.92
4711951501;2015-07-10T02:00+02:00;969.92
4711951501;2015-07-11T02:00+02:00;1697.92
4987464748;2015-07-12T02:00+02:00;0
4711951501;2015-07-13T02:00+02:00;1410.02
4711951501;2015-07-14T02:00+02:00;1481.12
4711951501;2015-07-15T02:00+02:00;1456.01
4711951501;2015-07-16T02:00+02:00;1494.21
4711951501;2015-07-17T02:00+02:00;1318.15
4711951501;2015-07-18T02:00+02:00;1368.26
4711951501;2015-07-19T02:00+02:00;1610.34
4711951501;2015-07-20T02:00+02:00;1394.24
4711951501;2015-07-21T02:00+02:00;669.94
4711951501;2015-07-22T02:00+02:00;687.89
4711951501;2015-07-23T02:00+02:00;687.74
4711951501;2015-07-24T02:00+02:00;590.54
4711951501;2015-07-25T02:00+02:00;591.51
4711951501;2015-07-26T02:00+02:00;558.30
4711951501;2015-07-27T02:00+02:00;964.36
4711951501;2015-07-28T02:00+02:00;884.35
